src/remoted/secure.c

```diff
@@ -689,39 +695,76 @@ STATIC void HandleSecureMessage(const message_t *message, int *wdb_sock) {
 
     key_unlock();
 
-    // Print first letter of the message
-    if (tmp_msg[0] == SYSCOLLECTOR_MQ) {
+
+
+    /* If we can't send the message, try to connect to the
+    * socket again. If it not exit.
+    */
+    if (SendMSG(logr.m_queue, tmp_msg, srcmsg, SECURE_MQ) < 0) {
+        merror(QUEUE_ERROR, DEFAULTQUEUE, strerror(errno));
+
+        // Try to reconnect infinitely
+        logr.m_queue = StartMQ(DEFAULTQUEUE, WRITE, INFINITE_OPENQ_ATTEMPTS);
+
+        minfo("Successfully reconnected to '%s'", DEFAULTQUEUE);
+
+        if (SendMSG(logr.m_queue, tmp_msg, srcmsg, SECURE_MQ) < 0) {
+            // Something went wrong sending a message after an immediate reconnection...
+            merror(QUEUE_ERROR, DEFAULTQUEUE, strerror(errno));
+        } else {
+            rem_inc_recv_evt(agentid_str);
+        }
+    } else {
+        rem_inc_recv_evt(agentid_str);
+    }
+
+    // Both syscollector delta and sync messages are sent to the router
+    if (tmp_msg[0] == SYSCOLLECTOR_MQ || tmp_msg[0] == DBSYNC_MQ) {
+        w_mutex_lock(&router_provider_mutex);
         if (!router_provider_handle) {
             if (router_provider_handle = router_provider_create("syscollector"), !router_provider_handle)
             {
                 os_free(agentid_str);
+                mdebug2("Failed to create router provider handle.");
+                w_mutex_unlock(&router_provider_mutex);
                 return;
             }
         }
 
-        router_provider_send(router_provider_handle, tmp_msg, msg_length);
-
-    } else {
-        /* If we can't send the message, try to connect to the
-        * socket again. If it not exit.
-        */
-        if (SendMSG(logr.m_queue, tmp_msg, srcmsg, SECURE_MQ) < 0) {
-            merror(QUEUE_ERROR, DEFAULTQUEUE, strerror(errno));
-
-            // Try to reconnect infinitely
-            logr.m_queue = StartMQ(DEFAULTQUEUE, WRITE, INFINITE_OPENQ_ATTEMPTS);
-
-            minfo("Successfully reconnected to '%s'", DEFAULTQUEUE);
+        // We only send the JSON part of the message
+        char* msg_start = strchr(tmp_msg, '{');
+        if(msg_start) {
```

Comment:
This conditional could be true even if msg_start is found outside the buffer. Maybe msg_start could be compared to be lower than the end position of the buffer (buffer + OS_MAXSTR + 1).

Reply:
Now that `strchr()` won't be used, this section was updated. Thank you.